### PR TITLE
[Debugger] Add support for remote call action evaluations

### LINKED ISSUE #31094: [Debugger Expression Evaluation] Add remote call action evaluation support
**Description:**
Debugger evaluation engine implementation should be improved to support remote method call actions.

**Describe your problem(s)**

**Describe your solution(s)**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by NipunaRanasinghe**
## Purpose
This PR 
- adds debugger expression evaluation support for remote call actions.
- adds integration tests to cover the related scenarios. 
- fixes #31094.
- fixes https://github.com/ballerina-platform/ballerina-lang/issues/27745.

## Remarks
We have found an issue where the evaluator returns `nil` when evaluating remote method call actions, which consumes multiple Ballerina strands. https://github.com/ballerina-platform/ballerina-lang/issues/31096 is created to track the progress.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [x] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [x] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/action/RemoteMethodCallActionEvaluator.java [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+    private Value invokeRemoteMethod(BVariable resultVar) throws EvaluationException {
+
+        ClassDefinitionResolver classDefResolver = new ClassDefinitionResolver(context);
+        String className = resultVar.getDapVariable().getValue();
+        Optional<ClassSymbol> classDef = classDefResolver.findBalClassDefWithinModule(className);
+        if (classDef.isEmpty()) {
+            // Resolves the JNI signature to see if the the object/class is defined with a dependency module.
+            String signature = resultVar.getJvmValue().type().signature();
+            if (!signature.startsWith("L")) {
```
Comment (praveennadarajah) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/action/RemoteMethodCallActionEvaluator.java:98]: What are the possible values can `signature` get here?
Do the remote method signature always starts with `L`?

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/action/RemoteMethodCallActionEvaluator.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+                throw new EvaluationException(String.format(EvaluationExceptionKind.CLASS_NOT_FOUND.getString(),
+                        className));
+            }
+
+            String[] signatureParts = signature.substring(1).split("/");
```
Comment (praveennadarajah) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/action/RemoteMethodCallActionEvaluator.java:103]: Will splitting the signature with forward slash `/` work with Windows?
I believe the signature will contain backward slashes in Windows(like how the path separator is different from Windows and Linux).

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/expression/MethodCallExpressionEvaluator.java [hunk 3]
```diff
@@ -117,8 +128,27 @@ public BExpressionValue evaluate() throws EvaluationException {
...
         boolean isFoundObjectMethod = false;
         try {
+            ClassDefinitionResolver classDefResolver = new ClassDefinitionResolver(context);
             String className = resultVar.getDapVariable().getValue();
-            Optional<ClassSymbol> classDef = findClassDefWithinModule(className);
+            Optional<ClassSymbol> classDef = classDefResolver.findBalClassDefWithinModule(className);
+            if (classDef.isEmpty()) {
+                // Resolves the JNI signature to see if the the object/class is defined with a dependency module.
+                String signature = resultVar.getJvmValue().type().signature();
+                if (!signature.startsWith("L")) {
```
Comment (praveennadarajah) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/expression/MethodCallExpressionEvaluator.java:137]: Same as https://github.com/ballerina-platform/ballerina-lang/pull/31097/files#r649648333

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/expression/MethodCallExpressionEvaluator.java [hunk 4]
```diff
@@ -117,8 +128,27 @@ public BExpressionValue evaluate() throws EvaluationException {
...
+                    throw new EvaluationException(String.format(EvaluationExceptionKind.CLASS_NOT_FOUND.getString(),
+                            className));
+                }
+
+                String[] signatureParts = signature.substring(1).split("/");
```
Comment (praveennadarajah) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/expression/MethodCallExpressionEvaluator.java:142]: Same as https://github.com/ballerina-platform/ballerina-lang/pull/31097/files#r649650492

File: tests/jballerina-debugger-integration-test/src/test/java/org/ballerinalang/debugger/test/adapter/evaluation/ExpressionEvaluationBaseTest.java [hunk 5]
```diff
@@ -212,4 +213,7 @@ protected void prepareForEvaluation() throws BallerinaTestException {
 
     // 33. XML navigation expressions
     public abstract void xmlNavigationEvaluationTest() throws BallerinaTestException;
+
+    // 33. XML navigation expressions
```
Comment (praveennadarajah) [inline tests/jballerina-debugger-integration-test/src/test/java/org/ballerinalang/debugger/test/adapter/evaluation/ExpressionEvaluationBaseTest.java:217]:
```suggestion
    // 34. Remote call action expressions
```

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/action/RemoteMethodCallActionEvaluator.java (hunk 1 above)
Comment (NipunaRanasinghe) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/action/RemoteMethodCallActionEvaluator.java:98]: Please refer JNI signature documentation. http://journals.ecs.soton.ac.uk/java/tutorial/native1.1/implementing/method.html#:~:text=Forming%20the%20Method%20Name%20and,name%20in%20UTF%2D8%20form.

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/action/RemoteMethodCallActionEvaluator.java (hunk 2 above)
Comment (NipunaRanasinghe) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/action/RemoteMethodCallActionEvaluator.java:103]: JNI signatures are not OS-dependent (as it's not a path).

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/expression/MethodCallExpressionEvaluator.java (hunk 3 above)
Comment (NipunaRanasinghe) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/expression/MethodCallExpressionEvaluator.java:137]: https://github.com/ballerina-platform/ballerina-lang/pull/31097#discussion_r649648333

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/expression/MethodCallExpressionEvaluator.java (hunk 4 above)
Comment (NipunaRanasinghe) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/expression/MethodCallExpressionEvaluator.java:142]: https://github.com/ballerina-platform/ballerina-lang/pull/31097#discussion_r649660977

File: tests/jballerina-debugger-integration-test/src/test/java/org/ballerinalang/debugger/test/adapter/evaluation/ExpressionEvaluationBaseTest.java (hunk 5 above)
Comment (NipunaRanasinghe) [reply tests/jballerina-debugger-integration-test/src/test/java/org/ballerinalang/debugger/test/adapter/evaluation/ExpressionEvaluationBaseTest.java]: Fixed with https://github.com/ballerina-platform/ballerina-lang/pull/31097#discussion_r649660977

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/action/RemoteMethodCallActionEvaluator.java (hunk 1 above)
Comment (NipunaRanasinghe) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/action/RemoteMethodCallActionEvaluator.java:98]: Screenshot from 2021-06-11 08-47-00

Comment (praveennadarajah) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/action/RemoteMethodCallActionEvaluator.java:98]: Ack!

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/action/RemoteMethodCallActionEvaluator.java (hunk 2 above)
Comment (praveennadarajah) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/action/RemoteMethodCallActionEvaluator.java:103]: Ack!

File: tests/jballerina-debugger-integration-test/src/test/java/org/ballerinalang/debugger/test/adapter/evaluation/ExpressionEvaluationTest.java [hunk 6]
```diff
@@ -700,6 +700,16 @@ public void xmlNavigationEvaluationTest() throws BallerinaTestException {
         // Todo
     }
 
+    @Override
+    @Test
+    public void remoteCallActionEvaluationTest() throws BallerinaTestException {
+        debugTestRunner.assertExpression(context, String.format("%s->getName(\"John\")", CLIENT_OBJECT_VAR), "John",
+                "string");
+        // Todo - Enable after fixing https://github.com/ballerina-platform/ballerina-lang/issues/31096
+        // debugTestRunner.assertExpression(context, String.format("%s->getTotal(78,90)", CLIENT_OBJECT_VAR), "168",
```
Comment (praveennadarajah) [inline tests/jballerina-debugger-integration-test/src/test/java/org/ballerinalang/debugger/test/adapter/evaluation/ExpressionEvaluationTest.java:709]:
```suggestion
        // debugTestRunner.assertExpression(context, String.format("%s-> getTotalMarks(78,90)", CLIENT_OBJECT_VAR), "168",
```

Comment (NipunaRanasinghe) [reply tests/jballerina-debugger-integration-test/src/test/java/org/ballerinalang/debugger/test/adapter/evaluation/ExpressionEvaluationTest.java]: Fixed with https://github.com/ballerina-platform/ballerina-lang/issues/27745

**Review by praveennadarajah — APPROVED** [review]

**Pull request merged**